### PR TITLE
Cypher engine and querying app

### PR DESCRIPTION
**Pull request by l-hoang**
Few years worth of work/commits in this PR re: the Cypher querying engine.

The main differences from the current master branch files aside from completely new files are the following:

- LC_CSR(-CSC)_Graph have a few degree getting functions, and CSR has a direct load-into-memory function as well.
- Pangolin libs have a bunch of ifdefs added by Roshan to facilitate turning on/off certain features in the QueryGraph.

New things include:
- libquery, where the Cypher compiler (parses cypher and changes to a format Galois engine understands) and querying engine reside.
- New submodule for a cypher parser used by the cypher compiler
- querying subdirectory in lonestar which currently contains an LDBC reader and querytest with querytest being the main addition in this PR (allows running queries on Katana Form graphs)
- changes to the result checker script to work with strings (used in querytest tests)

Note that the code will eventually need a modernization of sorts as it was written C-style to have it work with the use-case it was originally intended for. The namespacing and file layout could also use an improvement eventually. The point of this PR is to get this code out for people to use/look at ASAP.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -22,6 +22,7 @@ endif()
 
 ###### Options (alternatively pass as options to cmake -DName=Value) ######
 ###### Distributed-heterogeneous features ######
+set(GALOIS_ENABLE_CYPHER OFF CACHE BOOL "Enable Cypher/query graph library support")
```
Comment (roshandathathri) [inline CMakeLists.txt:25]: Don't think this is required but this is fine for now because enabling this changes Pangolin (library and apps).

File: libgalois/include/galois/graphs/LC_CSR_CSC_Graph.h [hunk 2]
```diff
@@ -314,6 +314,15 @@ class LC_CSR_CSC_Graph
                                          in_edge_end(N, mflag));
   }
 
+  /**
+   * @param N node to get in-degree for
+   * @param mflag how safe the acquire should be
+   * @returns In-degree of node N
+   */
+  auto in_degree(GraphNode N) const {
```
Comment (roshandathathri) [inline libgalois/include/galois/graphs/LC_CSR_CSC_Graph.h:322]: Please remove this function. getInDegree() already exists (I added it recently to master).

File: libgalois/include/galois/graphs/LC_CSR_Graph.h [hunk 3]
```diff
@@ -459,6 +459,15 @@ class LC_CSR_Graph :
                                          edge_end(N, mflag));
   }
 
+  /**
+   * @param N node to get degree for
+   * @param mflag how safe the acquire should be
+   * @returns Degree of node N
+   */
+  auto degree(GraphNode N) const {
```
Comment (roshandathathri) [inline libgalois/include/galois/graphs/LC_CSR_Graph.h:467]: Please remove this function. getDegree() already exists (I added it recently to master).

File: libgalois/include/galois/graphs/LC_CSR_Graph.h [hunk 4]
```diff
@@ -796,10 +805,58 @@ class LC_CSR_Graph :
    */
   const EdgeIndData& getEdgePrefixSum() const { return edgeIndData; }
 
+  //! Read topology of graph directly into allocated arrays (no edge data)
+  void readGraphTopology(int fd, size_t numNodes, size_t numEdges) {
```
Comment (roshandathathri) [inline libgalois/include/galois/graphs/LC_CSR_Graph.h:809]: Why is this needed? Please drop it if it is not used.

File: libgalois/include/galois/graphs/LC_CSR_Labeled_Graph.h [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+  auto degree(GraphNode N, const EdgeTy& data) const {
+    return std::distance(raw_begin(N, data), raw_end(N, data));
+  }
+
+  /**
+   * @param N node to get degree for
+   * @param mflag how safe the acquire should be
+   * @returns Degree of node N
+   */
+  auto degree(GraphNode N) const {
```
Comment (roshandathathri) [inline libgalois/include/galois/graphs/LC_CSR_Labeled_Graph.h:295]: Please drop this function

File: libgalois/include/galois/graphs/LC_CSR_Labeled_Graph.h [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+                                         edge_end(N, data, mflag));
+  }
+
+  /**
+   * @param N node to get degree for
+   * @param data label to get degree of
+   * @param mflag how safe the acquire should be
+   * @returns Degree of node N
+   */
+  auto degree(GraphNode N, const EdgeTy& data) const {
```
Comment (roshandathathri) [inline libgalois/include/galois/graphs/LC_CSR_Labeled_Graph.h:286]: degree -> getDegree

File: libgalois/include/galois/graphs/LC_CSR_Labeled_Graph.h [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+                                         in_edge_end(N, data, mflag));
+  }
+
+  /**
+   * @param N node to get in-degree for
+   * @param data label to get in-degree of
+   * @param mflag how safe the acquire should be
+   * @returns In-degree of node N
+   */
+  auto in_degree(GraphNode N, const EdgeTy& data) const {
```
Comment (roshandathathri) [inline libgalois/include/galois/graphs/LC_CSR_Labeled_Graph.h:377]: in_degree -> getInDegree

File: libgalois/include/galois/graphs/LC_CSR_Labeled_Graph.h [hunk 8]
```diff
@@ -0,0 +1,510 @@
...
+    return std::distance(in_raw_begin(N, data), in_raw_end(N, data));
+  }
+
+  /**
+   * @param N node to get in-degree for
+   * @param mflag how safe the acquire should be
+   * @returns In-degree of node N
+   */
+  auto in_degree(GraphNode N) const {
```
Comment (roshandathathri) [inline libgalois/include/galois/graphs/LC_CSR_Labeled_Graph.h:386]: Please drop this function

File: CMakeLists.txt (hunk 1 above)
Comment (l-hoang) [reply CMakeLists.txt:25]: yes, i think pangolin apps break if this is turn on as you mentioned in the issue you just opened

**Comment by l-hoang** [comment]
@ddn0 Should all 300+ commits be quashed into one? I think there may be some worth in revisiting older commits from this line of work as there are some older features in the history that we may want to bring back at some point (we've tried some interesting things over the years that aren't currently supported in this current state), but having a single neat commit is also a huge plus. I don't really have any *strong* opinion either way anyways, but I lean towards keeping the history.

What do you think @roshandathathri ?

**Comment by roshandathathri** [comment]
> Should all 300+ commits be quashed into one?

I'm opposed to it. I want the revision history.

File: libgalois/include/galois/graphs/LC_CSR_CSC_Graph.h [hunk 9]
```diff
@@ -430,6 +430,24 @@ class LC_CSR_CSC_Graph
...
+   * Returns in-degrees in a vector; useful if in-degrees need to be accessed
+   * quickly (1 memory access instead of 2 from subtracting begin and end).
+   */
+  auto countInDegrees() {
+    gstl::Vector<uint32_t> savedInDegrees(BaseGraph::numNodes);
+    galois::do_all(
+        galois::iterate(this->begin(), this->end()),
+        [&](unsigned v) {
+          savedInDegrees[v] =
+              std::distance(this->in_edge_begin(v), this->in_edge_end(v));
```
Comment (roshandathathri) [inline libgalois/include/galois/graphs/LC_CSR_CSC_Graph.h:446]: Shouldn't this be calling getInDegree()? 

Calling non-raw begin and end prevents the function from being const.

File: libgalois/include/galois/graphs/LC_CSR_CSC_Graph.h [hunk 10]
```diff
@@ -430,6 +430,24 @@ class LC_CSR_CSC_Graph
...
     constructIncomingEdges();
   }
+
+  ////////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Returns in-degrees in a vector; useful if in-degrees need to be accessed
+   * quickly (1 memory access instead of 2 from subtracting begin and end).
+   */
+  auto countInDegrees() {
```
Comment (roshandathathri) [inline libgalois/include/galois/graphs/LC_CSR_CSC_Graph.h:440]: { -> const {

**Comment by ddn0** [comment]
> Should all 300+ commits be quashed into one?... I don't really have any strong opinion either way anyways, but I lean towards keeping the history.

Without a culture of common code review and continuous testing, the value of any commit history is low for the people who weren't intimately involved with the development process. So, I'm not in a good position to value the history as you have it.

What I do care about is making sure our current state has well tested code paths, clean interfaces, etc. If there is any obvious clean up you can do now, I'd suggest you do it now so those who work on this code after you have the benefit your wisdom rather than the history of your actions.

**Comment by roshandathathri** [comment]
> those who work on this code after you have the benefit your wisdom rather than the history of your actions

Poetic. Wow!

**Comment by l-hoang** [comment]
FYI: I have another commit in the works that is meant to remove some dead code and does some streamlining, so @remaining reviewers you may want to wait (will be ready sometime tomorrow afternoon at the latest I hope).

**Comment by l-hoang** [comment]
Thanks everyone; will merge after this cleanup commit I'm about to push in + after CI passes for those. The only person who really needs to check the cleanup commit will be @roshandathathri (more details on what the commit has will be provided once I push).

File: lonestar/CMakeLists.txt [hunk 11]
```diff
@@ -215,6 +215,10 @@ if(GALOIS_ENABLE_DIST)
   endfunction()
 endif()
 
+if(GALOIS_ENABLE_CYPHER)
```
Comment (ddn0) [inline lonestar/CMakeLists.txt:218]: Can we just enable cypher by default?

(You know me and my quest to reduce build configurations :D )

File: lonestar/querying/cpu/querytest/querytest.cpp [hunk 12]
```diff
@@ -0,0 +1,191 @@
+#include "galois/Galois.h"
```
Comment (ddn0) [inline lonestar/querying/cpu/querytest/querytest.cpp:1]: Should I file a ticket to integrate some query tests into the CI?

Comment (l-hoang) [reply lonestar/querying/cpu/querytest/querytest.cpp:1]: Yes; they're actually already there. I'll just need to upload the inputs/outputs to the s3 bucket and fix a few assumptions probably.  That can be taken care of separately from this PR as well.

File: lonestar/CMakeLists.txt (hunk 11 above)
Comment (roshandathathri) [reply lonestar/CMakeLists.txt:218]: I agree but we cannot do this as of now because enabling cypher changes/breaks Pangolin and its apps. I'll try to fix that soon so that we can enable it by default but it won't be in this PR.

File: lonestar/querying/cpu/querytest/querytest.cpp (hunk 12 above)
Comment (roshandathathri) [reply lonestar/querying/cpu/querytest/querytest.cpp:1]: Yes, I think we can do that once cypher is enabled by default (see my comment above).

File: lonestar/CMakeLists.txt (hunk 11 above)
Comment (l-hoang) [reply lonestar/CMakeLists.txt:218]: There was a weird interdependency between this code and Pangolin in which enabling cypher broke things. I think I may have fixed it some time ago; I'll get back to you in a bit.

If everything builds fine with CYPHER enabled then yes, I have no problem with this; reducing build configs is good.

Comment (l-hoang) [reply lonestar/CMakeLists.txt:218]: @roshandathathri This *might* be fixed now, will check in a bit.

Comment (roshandathathri) [reply lonestar/CMakeLists.txt:218]: It's not just about building. Pangolin should run correctly without performance degradation. I would suggest not to attempt that in this PR.

Comment (l-hoang) [reply lonestar/CMakeLists.txt:218]: Sure thing.

Just checked and it *does* build and run successfully with ctest (not sure if ctest for shared mem apps check for correctness) for what it's worth. We can check performance later.

**Comment by l-hoang** [comment]
Set of commits pushed.

- removed dead/unused code/functions
- added more documentation
- **Cypher enabled by default**
- removed support for older way of loading graphs (maintaining support would be a pain in the future now that we're moving to KatanaForm)

Note that Cypher is enabled by default: I removed all coupling of Pangolin and libquery as the headers we're actually using from Pangolin don't touch the files that had the #ifdefs in question. libpangolin is pretty much in line with the master branch now.

File: lonestar/CMakeLists.txt (hunk 11 above)
Comment (l-hoang) [reply lonestar/CMakeLists.txt]: Note I'm making it build by default now because Pangolin has been completely decoupled from the querying library (as it should be).

Comment (roshandathathri) [reply lonestar/CMakeLists.txt]: > not sure if ctest for shared mem apps check for correctness

No, it doesn't. Only distributed analytics apps verify correctness.

> Note I'm making it build by default now because Pangolin has been completely decoupled from the querying library (as it should be).

Great, thanks!

File: libquery/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+  ${CMAKE_SOURCE_DIR}/libgalois/include
+  ${CMAKE_CURRENT_SOURCE_DIR}/include
+)
+
+add_dependencies(galois_query cypher_parser)
+target_link_libraries(galois_query ${Boost_LIBS} lonestar pangolin ${CYPHER_PARSER_LIBRARY})
+target_include_directories(galois_query PUBLIC ${CYPHER_PARSER_INCLUDE})
+
+if (CMAKE_HAVE_PTHREAD_H)
+   add_definitions(-DGALOIS_HAVE_PTHREAD)
```
Comment (roshandathathri) [inline libquery/CMakeLists.txt:36]: Why is this required?

File: libquery/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+   target_link_libraries(galois_query Threads::Threads)
+endif()
+
+find_package(NUMA)
+if(NUMA_FOUND)
+  add_definitions(-DGALOIS_USE_NUMA)
```
Comment (roshandathathri) [inline libquery/CMakeLists.txt:42]: Shouldn't this be done in the common cmake? Why is this in libquery?

File: lonestar/querying/cpu/querytest/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,20 @@
+add_executable(querytest querytest.cpp)
+target_link_libraries(querytest galois_query)
+# TODO shouldn't need to use this define; remove it in the future
+target_compile_definitions(querytest PRIVATE GALOIS_USE_QUERY_GRAPH_TYPE=1)
```
Comment (roshandathathri) [inline lonestar/querying/cpu/querytest/CMakeLists.txt:4]: No longer used. Drop it.

File: libquery/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+  src/AttributedGraph.cpp
+  src/CypherCompiler.cpp
+  src/Matching.cpp
+  src/SubgraphQuery.cpp
+  src/IndexableChunkedArray.cpp
+  src/KFHandler.cpp
+)
+
+# TODO avoid this
+target_compile_definitions(galois_query PRIVATE GALOIS_USE_QUERY_GRAPH_TYPE=1)
```
Comment (roshandathathri) [inline libquery/CMakeLists.txt:23]: No longer used. Drop this.

File: libquery/CMakeLists.txt (hunk 13 above)
Comment (l-hoang) [reply libquery/CMakeLists.txt:36]: I have no clue actually. I think it may have been a relic for when it was interfacing with the Python code and being built separately from libgalois.

File: libquery/CMakeLists.txt (hunk 14 above)
Comment (l-hoang) [reply libquery/CMakeLists.txt:42]: See reply for PTHREAD.

**Comment by l-hoang** [comment]
So the problem with making cypher default at the moment is that the CI workflow isn't applying the patch required for the cypher parser submodule to compile. I'll check the workflow and see how to do it.

**Comment by l-hoang** [comment]
Merging soon; feel free to make a last check if necessary.

The workflow things I had to add seem quite wordy/bad, so we'll have to fix that in a later PR.
